main/src/ca/uwaterloo/flix/language/phase/Regions.scala

```diff
@@ -421,16 +422,51 @@ object Regions {
     */
   private def checkType(tpe: Type, loc: SourceLocation)(implicit scope: List[Type.Var], flix: Flix): Validation[Unit, CompilationMessage] = {
     // Compute the region variables that escape.
-    val escapes = regionVarsOf(tpe) -- scope
+    val minned = TypeMinimization.minimizeType(tpe)
+    val regs = regionVarsOf(minned)
+    for (reg <- regs -- scope) {
+      if (essentialTo(reg, minned)) {
+        return TypeError.RegionVarEscapes(reg, minned, loc).toFailure
+      }
+    }
+    ().toSuccess
+  }
+
+  /**
+    * Returns true iff the type variable `tvar` is essential to the type `tpe`.
+    *
+    * A type variable is essential if its ascription has a bearing on the resulting value.
+    * For example, in the type `a and (not a)`, `a` is not essential since the result is always `false`.
+    */
+  def essentialTo(tvar: Type.Var, tpe: Type)(implicit flix: Flix): Boolean = {
+    if (!tpe.typeVars.contains(tvar)) {
+      // Case 1: The type variable is not present in the type. It cannot be essential.
+      false
+    } else {
+      // Case 2: The type variable is present in the type. Give it both values and see if the result changes.
+      val t0 = tpe.map {
+        case t if t == tvar => Type.False
+        case t => t
+      }
 
-    // Return an error if a region variable escapes.
-    if (escapes.nonEmpty) {
-      val rvar = escapes.head
-      return TypeError.RegionVarEscapes(rvar, tpe, loc).toFailure
+      val t1 = tpe.map {
+        case t if t == tvar => Type.True
+        case t => t
+      }
+
+      !sameType(t0, t1)
     }
+  }
 
-    // Otherwise return success.
-    ().toSuccess
+  /**
+    * Returns true iff the two types denote the same Boolean function, using the same variables.
+    */
+  private def sameType(t1: Type, t2: Type)(implicit flix: Flix): Boolean = {
```

Comment:
I was hoping to avoid extra complexity but I've hacked together a truth-tabley implementation to see what that does.

Reply:
The hacky truth-tabley implementation is no good since we are checking whether `r` is relevant to `t` where `t` is any kind of type, not just `Bool`.

I think we can do this by extracting all the boolean types from the type and seeing if `r` is relevant to them

Reply:
Its fun that I'm exactly looking at this case in the region proof